src/Interpreters/OptimizeDateFilterVisitor.cpp

```diff
@@ -0,0 +1,122 @@
+#include <Interpreters/OptimizeDateFilterVisitor.h>
+
+#include <Common/DateLUT.h>
+#include <Common/DateLUTImpl.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+
+
+namespace DB
+{
+
+ASTPtr generateOptimizedDateFilterAST(const String & comparator, const String & converter, const String & column, UInt64 compare_to)
+{
+    const DateLUTImpl & date_lut = DateLUT::instance();
+
+    if (converter != "toYear") return {};
+
+    UInt64 year = compare_to;
+    String from_date = date_lut.dateToString(date_lut.makeDayNum(year, 1, 1));
```

Comment:
maybe `start_date` / `end_date` will be clearer

Reply:
Exactly, revised accordingly. Thanks!